Separate individual static proxy variables from dynamic mesh variables

kyessenov opened the pull request.

rshriram commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -134,8 +135,18 @@ message ProxyMeshConfig {
-  // Path to the secrets used by the authentication policy.
-  string auth_certs_path = 101;
+  // Default proxy config for injected proxies
+  ProxyConfig default_config = 14;
```
What's this for? Need more comments here

kyessenov replied:
For kube-inject/initializer to apply settings to each new proxy.

costinm replied:
I guessed this - but please add it to the field comment. 
Can it be a different top level resource ? 
( i.e. the default proxy config will look like a normal proxy config )

rshriram commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -18,49 +18,15 @@ import "google/protobuf/duration.proto";
+// ProxyConfig defines variables for individual Envoy instances.
+message ProxyConfig {
+  // Path to the generated configuration file directory
+  string config_path = 1;
```
Config path or config file or both , to allow custom proxy configs ?

kyessenov replied:
This is not adding support for config file (no need to go through pilot for that).

rshriram commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -74,33 +40,68 @@ message ProxyMeshConfig {
+  string statsd_udp_address = 10;
+
+  // Port on which Envoy should listen for administrative commands.
+  int32 proxy_admin_port = 11;
```
How about we add xDS refresh interval (may be one per DS) and another config for inbound http port?

kyessenov replied:
eds, lds, cds refresh rates separately? OK.
Beware these cannot be changed dynamically. Only rds refresh rate.

rshriram replied:
Actually ignore my comment. I found that we have a global refresh rate config since 0.1.6.

ayj commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -134,8 +135,18 @@ message ProxyMeshConfig {
+  AuthPolicy auth_policy = 10;
+
+  // Polling interval for RDS (MUST BE >=1ms)
+  google.protobuf.Duration rds_refresh_delay = 11;
```
How is this different than the xDS refresh delays in ProxyConfig?

kyessenov replied:
For unknown reasons, RDS refresh rate is specified through LDS, so it's dynamically updatable.

ayj commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -74,33 +40,68 @@ message ProxyMeshConfig {
+  google.protobuf.Duration discovery_refresh_delay = 7;
+
+  // Address of the Zipkin service (e.g. _zipkin:9411_).
+  string zipkin_address = 8;
```
Why is zipkin_address duplicated in Proxy and Mesh config? Seems like ProxyConfig should only contain bootstrap config necessary to reach Pilot.

kyessenov replied:
For unknown reasons, parts of zipkin are static and parts are dynamic. This is the ugliness in xDS leaking to our proto.

ayj commented:
The PR includes static vs. dynamic distinction. It would be good to document the same in the proto itself.

costinm commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -18,49 +18,15 @@ import "google/protobuf/duration.proto";
 
```
Since today is code freeze date - and this while good doesn't seem critical change - can you keep the old config
around and add code to migrate/copy to the new one ? (i.e. follow standard process for API or config changes - 
deprecate old, migrate to new, remove old in next release). It's likely we have tests or configs using old format, this
may result in lots of eng spending time.

kyessenov replied:
There are only two places mesh configs are used - in pilot (fixed there in another PR) and in istio (which I'll have to fix).

costinm commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -18,49 +18,15 @@ import "google/protobuf/duration.proto";
-  // restart. MUST be >=1s (e.g., _1s/1m/1h_)
-  google.protobuf.Duration drain_duration = 7;
+  // Path to Envoy binary
+  string binary_path = 2;
```
Keep in mind for raw VM we use environment variables - cobra flags can be overriden by env, not sure if equivalent is possible for config files.

kyessenov replied:
`--config "binaryPath: $ENVOY"` works, no?

costinm commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -74,33 +40,68 @@ message ProxyMeshConfig {
+  // Connection timeout used by Envoy for supporting services. (MUST BE >=1ms)
+  google.protobuf.Duration connect_timeout = 9;
+
+  // IP Address and Port of a statsd UDP listener (e.g. _10.75.241.127:9125_).
```
We want to switch to statsd TCP if possible, maybe 'statsd_address=udp:.... | tcp:....'  ?
( so we don't have to change config again ) ? 

Or we'll add a statsd_tcp_address later ? What is closest to envoy config ?

kyessenov replied:
Need to follow up on the status of TCP support for statsd in envoy...

rshriram replied:
It already supports tcp statsd. We are waiting on mixer updating its image - which depends on tcp statsd to prometheus exporter

ldemailly replied:
we should have made that change before merging as anyway istio/api isn't automatically used by anyone
(pilot is still using Aug 28th version)

costinm commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -74,33 +40,68 @@ message ProxyMeshConfig {
+
+
+// ProxyMeshConfig defines mesh-wide variables shared by all Envoy instances in the
+// Istio service mesh.
```
Would be good to add comments on where it is expected to live.

Does mixer and CA read this file ?

kyessenov replied:
No, but they should. This will be another CRD, like a route rule.

kyessenov replied:
No but they should. This is meant to be another core config unit, a CRD.

costinm commented on proxy/v1/config/proxy_mesh.proto:
```diff
@@ -74,33 +40,68 @@ message ProxyMeshConfig {
-  string error_log_file = 15;
+  // Port on which Envoy should listen for incoming connections from
+  // other services.
+  int32 proxy_listen_port = 4;
```
Note that for each of those settings, we will likely need annotations/overrides in 0.3 - while in k8s ports are 'free', 
in raw VM some ports may be in use. 
That includes all ports, mixer address (VMs may need FQDN), and possibly auth policy. 

(maybe add it as TODO)

kyessenov replied:
This is precisely the point of separating. We can provide overrides per proxy over mesh-wide configs.

kyessenov replied:
Mixer is being configured remotely, not locally. So you need to define a "pod" notion for LDS to provide custom values.

rshriram commented:
Please fix the comment for config_path.

kyessenov commented:
I think I addressed most concerns. Let's merge this, and file separate issues to improve upon defaults, existing behavior. I have tried to limit the scope to static/dynamic distinction for now.
